jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullAndCacheBaseImageLayersStep.java

```diff
@@ -16,81 +16,44 @@
 
 package com.google.cloud.tools.jib.builder.steps;
 
-import com.google.cloud.tools.jib.async.AsyncDependencies;
-import com.google.cloud.tools.jib.async.AsyncStep;
-import com.google.cloud.tools.jib.async.NonBlockingSteps;
 import com.google.cloud.tools.jib.builder.ProgressEventDispatcher;
 import com.google.cloud.tools.jib.builder.TimerEventDispatcher;
-import com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.BaseImageWithAuthorization;
+import com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.ImageAndAuthorization;
 import com.google.cloud.tools.jib.configuration.BuildConfiguration;
 import com.google.cloud.tools.jib.image.Layer;
-import com.google.cloud.tools.jib.image.LayerPropertyNotFoundException;
 import com.google.common.collect.ImmutableList;
-import com.google.common.util.concurrent.ListenableFuture;
-import com.google.common.util.concurrent.ListeningExecutorService;
-import java.util.concurrent.Callable;
-import java.util.concurrent.ExecutionException;
+import java.util.ArrayList;
+import java.util.List;
 
 /** Pulls and caches the base image layers. */
-class PullAndCacheBaseImageLayersStep
-    implements AsyncStep<ImmutableList<PullAndCacheBaseImageLayerStep>>,
-        Callable<ImmutableList<PullAndCacheBaseImageLayerStep>> {
+// TODO: following the same pattern as "BuildAndCacheApplicationLayerStep", move the sole "makeList"
+// static into "PullAndCacheBaseImageLayerStep" and remove this class.
+class PullAndCacheBaseImageLayersStep {
 
-  private static final String DESCRIPTION = "Setting up base image caching";
+  private static final String DESCRIPTION = "Preparing base image layer pullers";
 
-  private final BuildConfiguration buildConfiguration;
-  private final ListeningExecutorService listeningExecutorService;
-  private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
-
-  private final PullBaseImageStep pullBaseImageStep;
-
-  private final ListenableFuture<ImmutableList<PullAndCacheBaseImageLayerStep>> listenableFuture;
-
-  PullAndCacheBaseImageLayersStep(
-      ListeningExecutorService listeningExecutorService,
+  static ImmutableList<PullAndCacheBaseImageLayerStep> makeList(
       BuildConfiguration buildConfiguration,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
-      PullBaseImageStep pullBaseImageStep) {
-    this.listeningExecutorService = listeningExecutorService;
-    this.buildConfiguration = buildConfiguration;
-    this.progressEventDispatcherFactory = progressEventDispatcherFactory;
-    this.pullBaseImageStep = pullBaseImageStep;
-
-    listenableFuture =
-        AsyncDependencies.using(listeningExecutorService)
-            .addStep(pullBaseImageStep)
-            .whenAllSucceed(this);
```

Comment:
Confirming: the core logic can start only after a base `Image` instance is created.

Knowing that, I put the core logic of "...LayersStep" (note the plural) into the "makeList()" static method instead of "call()", following the same pattern used in other steps that spawns sub-tasks: `makeList()` dynamically creates sub-tasks to pull each layer, and `call()` has the logic to download a layer. (To reduce code diff, I defer merging this class into `PullAndCacheBaseImageLayerStep` (note the singular) will be done in a follow-up.)